### PR TITLE
Wiktor/Migrate to RustCrypto

### PR DESCRIPTION
**Pull request by wiktor-k**
This is a continuation of #16 to see how far we can go :)

### CODE REVIEW

File: src/proto/message.rs [hunk 1]
```diff
@@ -1,134 +1,339 @@
...
+impl Decode for KeyConstraint {
+    type Error = Error;
+
+    fn decode(reader: &mut impl Reader) -> Result<Self> {
+        let constraint_type = u8::decode(reader)?;
+        // see: https://www.ietf.org/archive/id/draft-miller-ssh-agent-12.html#section-5.2
+        Ok(match constraint_type {
+            1 => KeyConstraint::Lifetime(u32::decode(reader)?),
+            2 => KeyConstraint::Confirm,
+            255 => KeyConstraint::Extension(String::decode(reader)?, Vec::<u8>::decode(reader)?),
```
Comment (baloo) [inline src/proto/message.rs:168]: lovely <3
This makes it much more flexible! I didn't realized we had the size of the vec in the payload and that we didn't require a proper serialization implementation of the content.

Comment (wiktor-k) [reply src/proto/message.rs:168]: I actually didn't test this yet but I found ssh-add has a destination constraint option so I'll capture some bytes for unit tests tomorrow (just like I did for SessionBind now).

In general I'll adjust key constraints extensions to work like regular extensions (see code below with Unparsed). That is it should be possible to deserialize them even if you define them outside of this crate. This gives the consumers maximum flexibility.

**Comment by wiktor-k** [comment]
Folks, a quick summary of what's going on:

Thanks to @jcspencer's work in #30 (that I've cherry-picked here) and [a bit of message capturing code](https://github.com/wiktor-k/ssh-agent-lib/pull/26/commits/0357ed38b0771bd1343fcbd02e72deeb4ff2d29b#diff-3bab5cd53a1040a4a531afe1501403786256e9dea8731dc20028c9beabddd2d3R43-R47) I've recorded a test session (`tests/sign-and-verify.sh`) and put all these requests in `tests/messages` (it occurred to me now that I've captured only requests, ugh, time to capture more!).

Then, I added a "roundtrip tests" thingy that takes all messages, decodes them and re-encodes and compares bytes. (The encoding part, once again, thanks to @jcspencer :bow: ).

I think this can severely simplify the recent tests that we have. First of all putting test artifacts in `tests/messages` automatically tests for decode/encode cycle so it's not necessary to do it twice. Second it's not necessary to hex-encode that... I hope you don't mind @baloo :sweat_smile: 

And the tests now can only check if the message that's decoded has correct stuff in appropriate places which should make them a bit more readable.

**Review by baloo — APPROVED** [review]
shipit shipit

**Comment by wiktor-k** [comment]
Okay folks... I'm going to clean up the commits a bit (too many `WIP`s from me :sweat_smile: ) and merge it. Once in `main` it's easier to stabilize and check what's still missing.

Thanks for your help and patience! :wave:

**Pull request merged**